x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -136,53 +146,260 @@ protected void readInternal(ByteBuffer b) throws IOException {
         final long position = getFilePointer() + this.offset;
         final int length = b.remaining();
 
-        int totalBytesRead = 0;
-        while (totalBytesRead < length) {
-            final long pos = position + totalBytesRead;
-            final int len = length - totalBytesRead;
-            int bytesRead = 0;
-            try {
-                final CacheFile cacheFile = getCacheFileSafe();
-                try (Releasable ignored = cacheFile.fileLock()) {
-                    final Tuple<Long, Long> rangeToWrite = computeRange(pos);
-                    final Tuple<Long, Long> rangeToRead = Tuple.tuple(pos, Math.min(pos + len, rangeToWrite.v2()));
-
-                    bytesRead = cacheFile.fetchAsync(rangeToWrite, rangeToRead, (channel) -> {
-                        final int read;
-                        if ((rangeToRead.v2() - rangeToRead.v1()) < b.remaining()) {
-                            final ByteBuffer duplicate = b.duplicate();
-                            duplicate.limit(duplicate.position() + Math.toIntExact(rangeToRead.v2() - rangeToRead.v1()));
-                            read = readCacheFile(channel, pos, duplicate);
-                            assert duplicate.position() <= b.limit();
-                            b.position(duplicate.position());
-                        } else {
-                            read = readCacheFile(channel, pos, b);
+        // We can detect that we're going to read the last 16 bytes (that contains the footer checksum) of the file. Such reads are often
+        // executed when opening a Directory and since we have the checksum in the snapshot metadata we can use it to fill the ByteBuffer.
+        if (length == CodecUtil.footerLength() && isClone == false && position == fileInfo.length() - length) {
+            if (readChecksumFromFileInfo(b)) {
+                logger.trace("read footer of file [{}] at position [{}], bypassing all caches", fileInfo.physicalName(), position);
+                return;
+            }
+            assert b.remaining() == length;
+        }
+
+        logger.trace("readInternal: read [{}-{}] ([{}] bytes) from [{}]", position, position + length, length, this);
+
+        final Tuple<Long, Long> indexCacheMiss; // null if not a miss
+
+        // We prefer to use the index cache if the recovery is not done yet
+        if (directory.isRecoveryDone() == false) {
```

Comment:
Do we still want to keep this check around? Given that we now always open a `CacheFile`, should we always go with that first, and then make use of this fallback instead?

Reply:
I've gone back and forth on this a few times, but now think this is a better idea, see b50d1589f983176a3ad597dafb1f05d3c82f9825. If we can serve the data from disk we do so; if we can't then we try the cache index, and finally fall back on the blob store, regardless of the phase of the shard's life.

In the case of no evictions it doesn't make much difference (we never have the data on disk during recovery) but I think this flow will work better for evicted data.